py-rse/py-scripting.Rmd

```diff
@@ -733,41 +905,102 @@ df.plot.scatter(x='word_frequency', y='inverse_rank',
 knitr::include_graphics("figures/py-rse-scripting/jane_eyre.png")
 ```
 
+## Summary {#py-rse-py-scripting-summary}
+
+Why is building a simple command-line tool so complex?
+One answer is that the conventions for command-line programs
+have evolved over several decades,
+so libraries like `argparse` must now support several different generations of option handling.
+Another is that the things we want to do genuinely *are* complex:
+read from either standard input or a list of files,
+display help when asked to,
+respect parameters that might not be there,
+and so on.
+As with many other things in programming (and life),
+everyone wishes it was simpler,
+but no one can agree on what to throw away.
+
+The good news is that this complexity is a fixed cost:
+our template for command-line tools can be re-used for programs
+that are much larger than the examples shown in this chapter.
+Making tools that behave in ways people expect
+greatly increases the chances that others will find them useful.
+
+## Exercises {#py-rse-py-scripting-exercises}
```

Comment:
A number of exercises here have the reader make edits to the scripts we've been developing (i.e. `countwords.py` and `plotcounts.py`). As it stands the edits to `plotcounts.py` carry forward to the next chapter (i.e. the next time we see that script in the lesson narrative the changes made in the exercises will be included), but not the edits to `countwords.py`. I guess I'm just flagging that we might need a way of indicating which changes carry forward and which don't... or we might decide that either none or all of the edits made in exercises carry forward. I'm kind of in favor of the former, because if none of the edits carry forward then you can read the whole book without doing the exercises (which let's face it, many readers will do) and it will all still make sense.

Reply:
:+1: